src/test/libraryfilesynctest.cpp

```diff
@@ -0,0 +1,701 @@
+            // for Ubuntu 20.04 where the time stamps sometimes do not
+            // progress as expected, i.e. the new time stamp might equal
+            // the old time stamp even though the file was modified.
+        } while (newLastModified <= oldLastModified);
+        // The new modification time must be strictly greater than
+        // the old modification time to establish valid preconditions
+        // for the tests! Since those time stamps are generated as
+        // side effects by the OS we need these ugly quirks here.
+        ASSERT_GT(newLastModified, oldLastModified);
+    }
+
+    void removeFile() const {
+        ASSERT_TRUE(m_fileInfo.toQFile().remove());
+    }
+
+  private:
+    const QTemporaryDir m_tempDir;
+    const mixxx::FileInfo m_fileInfo;
+};
+
+class SyncTrackMetadataConfigScope final {
+  public:
+    explicit SyncTrackMetadataConfigScope(
+            UserSettingsPointer pConfig)
+            : m_pConfig(std::move(pConfig)),
+              m_syncTrackMetadataBackup(
+                      m_pConfig->getValue(
+                              mixxx::library::prefs::kSyncTrackMetadataConfigKey,
+                              false)) {
+    }
+    ~SyncTrackMetadataConfigScope() {
+        // Restore config
+        m_pConfig->setValue(
+                mixxx::library::prefs::kSyncTrackMetadataConfigKey,
+                m_syncTrackMetadataBackup);
+    }
+
+    void setConfig(bool syncTrackMetadata) const {
+        m_pConfig->setValue(
+                mixxx::library::prefs::kSyncTrackMetadataConfigKey,
+                syncTrackMetadata);
+    }
+
+  private:
+    const UserSettingsPointer m_pConfig;
+    const bool m_syncTrackMetadataBackup;
+};
+
+} // namespace
+
+class LibraryFileSyncTest : public LibraryTest {
+  protected:
+    explicit LibraryFileSyncTest(const QFileInfo& testFile)
+            : m_tempFileSystem(testFile) {
+        const auto trackRef = TrackRef::fromFileInfo(m_tempFileSystem.fileInfo());
+        EXPECT_FALSE(trackCollectionManager()->getTrackByRef(trackRef));
+        const auto pTrack = getOrAddTrackByLocation(trackRef.getLocation());
+        EXPECT_TRUE(trackCollectionManager()->getTrackByRef(trackRef));
+        m_trackId = pTrack->getId();
+    }
+
+    TrackPointer loadTrack() const {
+        const auto pTrack = trackCollectionManager()->getTrackById(m_trackId);
+        DEBUG_ASSERT(pTrack);
+        return pTrack;
+    }
+
+    void saveModifiedTrack(TrackPointer&& pTrack) const {
+        ASSERT_NE(nullptr, pTrack);
+        ASSERT_EQ(1, pTrack.use_count());
+        const auto trackId = pTrack->getId();
+        ASSERT_TRUE(trackId.isValid());
+        ASSERT_TRUE(pTrack->isDirty());
+        // The file might be modified if track metadata is exported.
+        // Ensures that the resulting time stamp will be strictly
+        // greater than the current one.
+        sleepBeforeUpdatingFileLastModifiedTime();
```

Comment:
The sleepBeforeUpdatingFileLastModifiedTime here might be insufficient? 

The sleep belongs to the context of the test where a > detects a file write. The best would be to age the original file at the beginning of the outer function.  Alternatively you may to rename the function to a name that reveals this nature. 
saveModifiedTrackAndAdvanceFileTime();

Reply:
Saving a track is implicitly supposed to touch the file depending on the sync settings.

Consequently following your proposal I would have to rename it to saveModifiedTrackAndMaybeExportFileTagsAndAdvanceFileTime() or stuff even more implementation details into the function name. This is not helpful.